serverless/serverless
# Add Cyclic Object Serialization Support

## What did you implement:

Re-Closes #3509 



Allow for the serialization and deserialization of cyclic objects being written to file using json-decycle.  (in support of #4144)

## How did you implement it:


Add `json-decycle` and use and instance of it as the second parameter for JSON.stringify/JSON.parse for file writing and reading.

## How can we verify it:


A project with a `${self:}` reference should be sufficient to evoke the issue.  During serialization you would get the result:
```
[...] $SLS_DEBUG=* slsDev deploy
Serverless: Packaging service...
    Serverless: Excluding development dependencies...

    Type Error ---------------------------------------------

    Converting circular structure to JSON

For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.

    Stack Trace --------------------------------------------

    TypeError: Converting circular structure to JSON
at Object.stringify (native)
at writeFileSync (~/serverless/lib/utils/fs/writeFileSync.js:13:21)
at Utils.writeFileSync (~/serverless/lib/classes/Utils.js:45:12)
at AwsPackage.saveServiceState (~/serverless/lib/plugins/aws/package/lib/saveServiceState.js:39:27)
From previous event:
    at Object.aws:package:finalize:saveServiceState [as hook] (~/serverless/lib/plugins/aws/package/index.js:81:10)
at BbPromise.reduce (~/serverless/lib/classes/PluginManager.js:242:55)
From previous event:
    at PluginManager.invoke (~/serverless/lib/classes/PluginManager.js:242:22)
at PluginManager.spawn (~/serverless/lib/classes/PluginManager.js:254:17)
at AwsPackage.BbPromise.bind.then (~/serverless/lib/plugins/aws/package/index.js:69:51)
From previous event:
    at Object.package:finalize [as hook] (~/serverless/lib/plugins/aws/package/index.js:69:10)
at BbPromise.reduce (~/serverless/lib/classes/PluginManager.js:242:55)
From previous event:
    at PluginManager.invoke (~/serverless/lib/classes/PluginManager.js:242:22)
at PluginManager.spawn (~/serverless/lib/classes/PluginManager.js:254:17)
at Deploy.BbPromise.bind.then.then (~/serverless/lib/plugins/deploy/deploy.js:101:50)
From previous event:
    at Object.before:deploy:deploy [as hook] (~/serverless/lib/plugins/deploy/deploy.js:99:10)
at BbPromise.reduce (~/serverless/lib/classes/PluginManager.js:242:55)
From previous event:
    at PluginManager.invoke (~/serverless/lib/classes/PluginManager.js:242:22)
at PluginManager.run (~/serverless/lib/classes/PluginManager.js:261:17)
at variables.populateService.then (~/serverless/lib/Serverless.js:99:33)
at runCallback (timers.js:672:20)
at tryOnImmediate (timers.js:645:5)
at processImmediate [as _immediateCallback] (timers.js:617:5)
From previous event:
    at Serverless.run (~/serverless/lib/Serverless.js:86:74)
at serverless.init.then (~/serverless/bin/serverless:39:50)

Get Support --------------------------------------------
    Docs:          docs.serverless.com
Bugs:          github.com/serverless/serverless/issues
Forums:        forum.serverless.com
Chat:          gitter.im/serverless/serverless

Your Environment Information -----------------------------
    OS:                     darwin
Node Version:           6.11.0
Serverless Version:     1.20.2
```

## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
